### PR TITLE
feat(graphql): ability to update/query for appData

### PR DESCRIPTION
**Pull request by lmiller1990**
Make the GraphQL interface for access/updating local preferences more generic.

```gql
# query
{
  localSettings {
    preferences {
      isSpecsListOpen
    }
  }
}

# update
mutation M {
  # value is JSON.stringify({ k1: value1, k1: value2, ... })
  setPreferences(value: "{ \"isSpecsListOpen\": true }")
}
```

Typing the `setPreferences ` mutation might be good but it seems a bit complex, and might not make sense since a lot of the fields we cache will either go away, or change, so I just set it to take an object which is JSON.stringified. This is how the current save state API works.

It also improves the types for local data. I hope to make these more strict as we go.

image

I also added a missing feature which is "open config in editor" since we needed that and it will use a lot of this functionality (eg, we need to set the correct preferred editor if they don't have one). Here is a demo (shows that the `localSettings` feature is working correctly, too):

https://user-images.githubusercontent.com/19196536/143182237-fe064a72-3802-45c6-93ad-7c0984f237cd.mov

### CODE REVIEW

**Review by marktnoonan — CHANGES_REQUESTED** [review]
I think one change is needed for the default values for the preference to survive the initialization, otherwise this works great.

File: packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts [hunk 1]
```diff
@@ -294,51 +294,14 @@ export const mutation = mutationType({
...
-        return true
-      },
-    })
-
-    t.liveMutation('setPreferredEditorBinary', {
+    t.liveMutation('setPreferences', {
       type: 'Boolean',
       args: {
+        // Stringify JSON, eg JSON.stringify({ firstTimeOpening: Date.now() })
         value: nonNull(stringArg()),
```
[the remark this answers is not in the document]
Comment (lmiller1990) [reply packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts:309]: Updated, good catch

File: packages/types/src/preferences.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
+import type { Editor } from '.'
+
+export const defaultPreferences: AllowedState = {
```
[the remark this answers is not in the document]
Comment (lmiller1990) [reply packages/types/src/preferences.ts:3]: Good point... updated to set defaults in the same place as we load the settings. Local preferences will override defaults.

[the remark this answers is not in the document]
Comment (lmiller1990) [reply packages/types/src/preferences.ts:3]: https://github.com/cypress-io/cypress/pull/19082/commits/6b83406160a83a7ada344ff6123fd43bec018103

**Review by marktnoonan — APPROVED** [review]
Looks good, I pushed 1 commit for an out of date description.

**Pull request merged**